pkg/migrate/cstor/pool.go

```diff
@@ -0,0 +1,467 @@
+			return errors.Wrap(err, "failed to get operator version")
+		}
+		if vOperator.LessThan(v1110) {
+			return fmt.Errorf("cspc operator is in %s version, please upgrade it to 1.11.0 or above version",
+				pod.Labels["openebs.io/version"])
+		}
+	}
+	return nil
+}
+
+// Pool migrates the pool from SPC schema to CSPC schema
+func (c *CSPCMigrator) migrate(spcName string) error {
+	var err error
+	var migrated bool
+	c.SPCObj, migrated, err = c.getSPCWithMigrationStatus(spcName)
+	if migrated {
+		klog.Infof("spc %s is already migrated to cspc", spcName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	err = c.validateSPC()
+	if err != nil {
+		return errors.Wrapf(err, "failed to validate spc %s", spcName)
+	}
+	err = c.updateBDCLabels(spcName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update bdc labels for spc %s", spcName)
+	}
+	klog.Infof("Creating equivalent cspc for spc %s", spcName)
+	c.CSPCObj, err = c.generateCSPC()
+	if err != nil {
+		return err
+	}
+	err = c.updateBDCOwnerRef()
+	if err != nil {
+		return err
+	}
+	// List all cspi created with reconcile off
+	cspiList, err := c.OpenebsClientset.CstorV1().
+		CStorPoolInstances(c.OpenebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + c.CSPCObj.Name,
+		})
+	if err != nil {
+		return err
+	}
+
+	// For each cspi perform the migration from csp that present was on
+	// node on which cspi came up.
+	for _, cspiItem := range cspiList.Items {
+		cspiItem := cspiItem // pin it
+		cspiObj := &cspiItem
+		err = c.cspTocspi(cspiObj)
+		if err != nil {
+			return err
+		}
+	}
+	// Clean up old SPC resources after the migration is complete
+	err = spc.NewKubeClient().
+		Delete(spcName, &metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// validateSPC determines that if the spc is allowed to migrate or not.
+// If the max pool count does not match the number of csp for auto spc, or
+// the bd list in spc does not match bds from the csp migration should not be done.
+func (c *CSPCMigrator) validateSPC() error {
+	cspClient := csp.KubeClient()
+	cspList, err := cspClient.List(metav1.ListOptions{
+		LabelSelector: string(apis.StoragePoolClaimCPK) + "=" + c.SPCObj.Name,
+	})
+	if err != nil {
+		return err
+	}
+	if c.SPCObj.Spec.BlockDevices.BlockDeviceList == nil {
+		if c.SPCObj.Spec.MaxPools == nil {
+			return errors.Errorf("invalid spc %s neither has bdc list nor maxpools", c.SPCObj.Name)
+		}
+		if *c.SPCObj.Spec.MaxPools != len(cspList.Items) {
+			return errors.Errorf("maxpool count does not match csp count expected: %d got: %d",
+				*c.SPCObj.Spec.MaxPools, len(cspList.Items))
+		}
+		return nil
+	}
+	bdMap := map[string]int{}
+	for _, bdName := range c.SPCObj.Spec.BlockDevices.BlockDeviceList {
+		bdMap[bdName]++
+	}
+	for _, cspObj := range cspList.Items {
+		for _, rg := range cspObj.Spec.Group {
+			for _, bdObj := range rg.Item {
+				bdMap[bdObj.Name]++
+			}
+		}
+	}
```

Comment:
Better to add the unit tests for these , can be added in next PR

Reply:
Will add in upcoming PRs.